### PR TITLE
raft: don't emit unstable CommittedEntries

### PR DESCRIPTION
**Pull request by tbg**
Fixes https://github.com/etcd-io/etcd/issues/14370.

When run in a single-voter configuration, prior to this PR
raft would emit `HardState`s that would emit a proposed `Entry`
simultaneously in `CommittedEntries` and `Entries`.

To be correct, this requires users of the raft library to enforce an
ordering between appending to the log and notifying the client about
`CommittedEntries` also present in `Entries`. This was easy to miss.

Walk back this behavior to arrive at a simpler contract: what's
emitted in `CommittedEntries` is truly committed, i.e. present
in stable storage on a quorum of voters.

This in turn pessimizes the single-voter case: rather than fully
handling an `Entry` in just one `Ready`, now two are required,
and in particular one has to do extra work to save on allocations.

We accept this as a good tradeoff, since raft primarily serves
multi-voter configurations.

----

Suggested review plan:

- Look at `raft: don't emit unstable CommittedEntries` first to get an idea of the gist of the change
- Look at `raft: directly update leader in advance` for an alternative solution that "inlines" the relevant bits of `r.Step` instead.
-  The test suites pass with either solution, so the two solutions are functionally equivalent to the degree that we're exercising them.
- Look at `raft: add BenchmarkRawNode` and at the benchmark results in `raft: benchmark results`
- If there are substantial concerns and/or comments, stop here and sort them out with me.
- Then slog through the remaining commits, which are all going to be squashed before the merge, and fix up the tests one-by-one.
- Note: `raft: always mark leader as RecentActive` was included here since otherwise the outputs of the two possible solutions differ in sometimes failing to track the leader as recently active.

### CODE REVIEW

**Comment by serathius** [comment]
Rerun the e2e flake. This PR is still failing DCO, please run `git rebase origin/main --signoff` and push head with force `git push origin HEAD:raft-single-voter -f`.

File: raft/raft.go
Comment (nvanbenschoten) [inline raft/raft.go:650]: nit: no need to update `li`.

**Review by nvanbenschoten — COMMENTED** [review]
This change makes sense to me. It feels like the "proper" solution because it simplifies the `Ready` contract and avoids the need for special-casing of single-peer deployments. After this change, entries in the `CommittedEntries` slice are always committed.

I actually made very similar changes in a prototype (https://github.com/nvanbenschoten/etcd/commit/1d1fa32ec22715f6e21a6d6486f865356c3eb4fc) intended to address https://github.com/etcd-io/etcd/issues/12257 / https://github.com/cockroachdb/cockroach/issues/17500. These kinds of extensions will be easier to make if the Raft leader doesn't immediately consider an entry to be durable upon appending to its own unstable log (in `raft.appendEntry`). Things are cleaner if, instead, the durability status is maintained either implicitly in `raft.advance` or explicitly through a different (opt-in) API if clients want the flexibility to sync their log asynchronously.

From the perspective of CRDB's use of Raft, we weren't susceptible to the bug in #14370 because log application always came after log appending in the Ready processing state machine loop. However, we were aware of this idiosyncrasy and had to work around it when acking proposals ahead of application, see: https://github.com/cockroachdb/cockroach/blob/ce55e1b46604401f6085a6ece1355970da94600a/pkg/kv/kvserver/replica_raft.go#L834-L844

The comment there is interesting because it references a second case where `Entries` and `CommittedEntries` can overlap — that being when a follower in a multi-node replication group is catching up after falling behind. In those cases, the overlapping entries are already committed, so there's no durability risk from applying CommittedEntries before appending Entries (like `etcdserver` does). However, that still means that it's possible to have applied an entry that is not durably stored locally in the Raft log. I don't know if that could cause issues for etcd or whether entry application is idempotent.

File: raft/raft.go
Comment (ahrtr) [reply raft/raft.go:650]: +1

File: raft/storage.go
Comment (ahrtr) [inline raft/storage.go:91]: I do not see where `callStats` is used. Are you planning to use it in a separate PR?

File: raft/node_util_test.go
Comment (ahrtr) [inline raft/node_util_test.go:1]: 2022?

File: raft/storage.go
Comment (ahrtr) [reply raft/storage.go:91]: It seems also an unrelated change, so suggest to do it in a separate PR.

Comment (ahrtr) [reply raft/storage.go:91]: Got it, it's used in the Benchmark test cases.

File: raft/rawnode_test.go
Comment (ahrtr) [inline raft/rawnode_test.go:1141]:
```suggestion
	cases := []struct{
	    name string
	    peers []uint64
	 } {
	     {
	         name: "single-voter",
	         peers: []uint64{1},
	     },
	     {
	         name: "two-voters",
	         peers: []uint64{1, 2},
	     },
	     // You can easily add more cases here
	 }
	 
	 for _, tc := range cases {
	     b.Run(tc.name, func(b *testing.B) {
	         benchmarkRawNodeImpl(b, tc.peers)
	     }
	 }
```

File: raft/diff_test.go
Comment (ahrtr) [inline raft/diff_test.go:58]: Why do you change this?

File: raft/node_util_test.go
Comment (ahrtr) [inline raft/node_util_test.go:26]:
```suggestion
	*testing.T
```

Comment (ahrtr) [reply raft/node_util_test.go:26]: After this change, you can remove all the following methods (e.g. Debug(f), Error(f), Info(f), Warning(f), Fatal(f), Panic(f)

File: raft/node_util_test.go
Comment (ahrtr) [inline raft/node_util_test.go:98]: There is a potential goroutine leak. When the test case finishes, it may still run. We need to make sure gracefully shutdown for every goroutine.

File: raft/diff_test.go
Comment (ahrtr) [reply raft/diff_test.go:58]: I am curious why this PR is much bigger than https://github.com/etcd-io/etcd/pull/14411, since they are almost following the same logic ? I am not sure whether it's related to this change?

File: raft/node_test.go
Comment (ahrtr) [inline raft/node_test.go:168]: Why did you remove this case ?

**Comment by ahrtr** [comment]
The production code change looks simple & good, but the test code change is too big. This PR almost follows the same logic as https://github.com/etcd-io/etcd/pull/14411, but this PR is much bigger, even excluding the new benchmark test case. 

Another comment... Have you compared the performance as compared to the `main` branch using the [benchmark](https://github.com/etcd-io/etcd/tree/main/tools/benchmark) tool? I suggest to compare both single-node cluster and multi-node cluster.  FYI. https://github.com/etcd-io/etcd/pull/14394#issuecomment-1229606410 and https://github.com/etcd-io/etcd/pull/14394#issuecomment-1229673875

**Comment by tbg** [comment]
Thanks for the review, @ahrtr. I'm not responding to the inline comments yet since you seem to have general concerns about the size of the change.

I think it comes down to the discussion in https://github.com/etcd-io/etcd/issues/14370#issuecomment-1237747454.

You're fixing `*node`, I'm fixing `raft`. So I have to update more tests.
Also, you keep the old behavior for empty entries for reasons that aren't clear to me. I think this prevents you from having to adjust more tests.

> Usually empty entries will only be generated in multi-node cluster. Since multi-node clusters do not have this issue, so I think it makes sense to keep the optimization (update the pr.Match immediately after appending it).

In my book, a special case needs a compelling reason. Special cases add complexity, and this complexity has to be made up for by some benefit. I cannot see that benefit here.

Some of the tests were pretty confused about whether they wanted to make statements about entry append or entry commit or entry apply. This is because the problem we're fixing made these things all the same in a single-node cluster in the same cycle.

I made sure to update the tests to be more robust (especially the `TestNode.*` tests are pretty annoying when they fail, prior to my improvements) and to not get too crufty. That is some extra work, but again, this PR should be judged by its changes to production code, not the corresponding updates in tests.

> 2. If you remove the optimization (mentioned above) so as to keep the behavior consistent for both empty and non-empty entry, then likely there are more subtle edge cases you need to handle, accordingly it might further complicate the raft package; and they are exactly what I have been trying to avoid.

Could you elaborate on this? Two cases = more complexity. In this PR, the empty entry is just an entry; why make it more complex? There should be fewer subtle edge cases now - namely none at all - and that is what we are after. What am I missing?

> 3. Actually https://github.com/etcd-io/etcd/pull/14411 doesn't almost change anything on the raft protocol, which I think is better. But I agree it has some impact on the direct users on RawNode.

We can agree to disagree here - *raft is the thing that produces `Ready` and the semantics of `Ready` are what need to be fixed. We could fix it in `*RawNode` but `RawNode` is intentionally a thin interface wrapper around `*raft` (similar to `*node` is just a goroutine around `*RawNode`). Spreading implementation details between them is not going to make things clearer. We shouldn't start to bolt on logic outside of `*raft` because we're scared of changing `*raft`.

**Comment by ahrtr** [comment]
Thanks for the feedback. 

> Could you elaborate on this? Two cases = more complexity. In this PR, the empty entry is just an entry; why make it more complex? There should be fewer subtle edge cases now - namely none at all - and that is what we are after. What am I missing?

I just kept the legacy behavior on empty entry so as to keep the PR(14411) as simple&safe as possible, and it's the reason why the size of 14411 is much smaller than this one. I agree that regarding the empty entry as a special case isn't good, although it was consistent with the legacy behavior. 





> Some of the tests were pretty confused about whether they wanted to make statements about entry append or entry commit or entry apply. This is because the problem we're fixing made these things all the same in a single-node cluster in the same cycle.

Thanks for the clarification. 



> That is some extra work, but this PR should be judged by its changes to production code, not the corresponding updates in tests.

Partially agreed. We also need to make sure the quality of test code. The reason why I raised the concern on the size on the test code is just to understand the reason of the bigger size. Good to confirmed that one of the major reasons is due to the removal of the special case on empty entry.  

Overall looks good to me. Please,
1. Consider to resolve the inline minor comments;
2. Consider to squash the commits into a couple of commits if possible.

**Comment by tbg** [comment]
Thanks @ahrtr. I'm travelling this week, so there won't be activity on this PR until next week but I plan to return to it the week after. I am also planning to validate the perf impact of this change with the etcd suite as you suggested, and also against CRDB.

File: raft/diff_test.go
Comment (tbg) [reply raft/diff_test.go:58]: The various tests that are using this really wanted to see what's in the raft log. That this was called "committed" was a misnomer, the tests really didn't care whether it was committed or not. It just so happened that append==commit in the single-node case prior to this change.

File: raft/node_util_test.go
Comment (tbg) [reply raft/node_util_test.go:26]: No, we are implementing `raft.Logger` here, `*testing.T` does not implement that. It is true that some methods overlap (for example, `*testing.T` has a matching `Fatalf`) but semantically we don't necessarily want to call `t.Fatalf` just because a raft node calls `Fatalf`. I will leave this as is.

File: raft/node_util_test.go
Comment (tbg) [reply raft/node_util_test.go:98]: Good call, fixed.

File: raft/node_test.go
Comment (tbg) [reply raft/node_test.go:168]:
```
commit 0a6a94b5e7570f2876e677831964628e9df94272
Author: Tobias Grieger <tobias.b.grieger@gmail.com>
Date:   Tue Sep 6 17:26:13 2022 +0200

    remove TestNodeReadIndex

    This is tested directly at the level of `RawNode` in
    `TestRawNodeReadIndex`. `*node` is a thin wrapper around `RawNode` so
    this is sufficient.

    The reason to remove the test is that it now incurs data races
    since it's not possible to adjust the `readStates` and `step`
    fields while the node is running, and there is no primitive
    to synchronize with its goroutine. This could all be fixed
    but it's not worth it.

    Signed-off-by: Tobias Grieger <tobias.b.grieger@gmail.com>
```

**Comment by tbg** [comment]
Addressed the first round of comments, see the fixup commit. I also added an (empty) commit describing the results for [this benchmark](https://github.com/etcd-io/etcd/pull/14394#issuecomment-1229606410).

TL;DR is that we're not regressing, in fact getting a little *faster*.

This is mildly surprising to me.
To be fair, I improved performance in this PR via some micro-optimizations which may result in fewer calls to `LastIndex`, `FirstIndex`, etc.
Also, the benchmark hopefully saturates the raft pipeline, meaning that even though we can't commit in a single round through raft any more we are doing lots of work in each raft cycle, i.e. we're fsyncing at about the same rate.
But I still didn't think we'd beat the main branch in latency, which seems to be happening, too, in both runs.

**Comment by ahrtr** [comment]
Overall looks good to me. 

Good to see the performance enhancement below, although it is an unrelated change. 
```
func releasePendingReadIndexMessages(r *raft) {
	if len(r.pendingReadIndexMessages) == 0 {
		// Fast path for the common case to avoid a call to storage.LastIndex()
		// via committedEntryInCurrentTerm.
		return
	}
	......
}
```

Please resolve the conflict.

**Comment by ahrtr** [comment]
Next steps:
1. After merging this PR, I am going to revert the PR https://github.com/etcd-io/etcd/pull/14400 on main branch;
2. Just as we discussed early, we will not backport this PR to release-3.5 and release-3.4;

Please let me know if you have any concerns or comments. cc @tbg @ptabor

**Review by ahrtr — APPROVED** [review]
Great work! Thank you @tbg

**Comment by ahrtr** [comment]
Since there is no any more comments or concerns from other maintainers, so let me merge this PR, otherwise we need to rebase this PR to resolve conflict from time to time.

**Pull request merged**
